src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,493 @@
+/*
+ * src/bin/pgcopydb/pgsql_timeline.c
+ *	 API for sending SQL commands about timelines to a PostgreSQL server
+ */
+
+#include "catalog.h"
+#include "file_utils.h"
+#include "log.h"
+#include "pg_utils.h"
+#include "pgsql_timeline.h"
+#include "pgsql_utils.h"
+#include "pgsql.h"
+
+
+typedef struct IdentifySystemResult
+{
+	char sqlstate[6];
+	bool parsedOk;
+	IdentifySystem *system;
+} IdentifySystemResult;
+
+
+typedef struct TimelineHistoryResult
+{
+	char sqlstate[6];
+	bool parsedOk;
+	char filename[MAXPGPATH];
+} TimelineHistoryResult;
+
+static void parseIdentifySystemResult(IdentifySystemResult *ctx, PGresult *result);
+static void parseTimelineHistoryResult(TimelineHistoryResult *context, PGresult *result,
+									   char *cdcPathDir);
+static bool writeTimelineHistoryFile(char *filename, char *content, char *cdcPathDir);
+
+
+/*
+ * pgsql_identify_system connects to the given pgsql client and issue the
+ * replication command IDENTIFY_SYSTEM. The pgsql connection string should
+ * contain the 'replication=1' parameter.
+ */
+bool
+pgsql_identify_system(PGSQL *pgsql, IdentifySystem *system, char *cdcPathDir)
+{
+	bool connIsOurs = pgsql->connection == NULL;
+
+	PGconn *connection = pgsql_open_connection(pgsql);
+	if (connection == NULL)
+	{
+		/* error message was logged in pgsql_open_connection */
+		return false;
+	}
+
+	/* extended query protocol not supported in a replication connection */
+	PGresult *result = PQexec(connection, "IDENTIFY_SYSTEM");
+
+	if (!is_response_ok(result))
+	{
+		log_error("Failed to IDENTIFY_SYSTEM: %s", PQerrorMessage(connection));
+		PQclear(result);
+		clear_results(pgsql);
+
+		PQfinish(connection);
+
+		return false;
+	}
+
+	IdentifySystemResult isContext = { { 0 }, false, system };
+
+	(void) parseIdentifySystemResult((void *) &isContext, result);
+
+	PQclear(result);
+	clear_results(pgsql);
+
+	log_sql("IDENTIFY_SYSTEM: timeline %d, xlogpos %s, systemid %" PRIu64,
+			system->timeline,
+			system->xlogpos,
+			system->identifier);
+
+	if (!isContext.parsedOk)
+	{
+		log_error("Failed to get result from IDENTIFY_SYSTEM");
+		PQfinish(connection);
+		return false;
+	}
+
+	/* while at it, we also run the TIMELINE_HISTORY command */
+	if (system->timeline > 1)
+	{
+		TimelineHistoryResult hContext = { 0 };
+
+		char sql[BUFSIZE] = { 0 };
+		sformat(sql, sizeof(sql), "TIMELINE_HISTORY %d", system->timeline);
+
+		result = PQexec(connection, sql);
+
+		if (!is_response_ok(result))
+		{
+			log_error("Failed to request TIMELINE_HISTORY: %s",
+					  PQerrorMessage(connection));
+			PQclear(result);
+			clear_results(pgsql);
+
+			PQfinish(connection);
+
+			return false;
+		}
+
+		(void) parseTimelineHistoryResult((void *) &hContext, result, cdcPathDir);
+
+		sformat(system->timelineHistoryFilename, MAXPGPATH,
+				"%s/%s", cdcPathDir, hContext.filename);
```

Comment:
The same `sformat` is done twice. Can we pass the system down to `parseTimelineHistoryResult` so that we build the timeline history absolute pathname only once?

Reply:
I agree that I should do this formatting only once. However, I'd rather not push `system` to `parseTimelineHistoryResult` as accessing that is not really necessary for parsing. Instead, I copy the string after this function completes.